### PR TITLE
Suggested changes to property override PR

### PR DESCRIPTION
**Pull request by nguerrera**
Sending this feedback on https://github.com/microsoft/cadl/pull/1248 using as a PR to your PR :)

Will comment on the rationale behind each change.

### CODE REVIEW

File: packages/compiler/core/checker.ts [hunk 1]
```diff
@@ -2226,21 +2224,20 @@ export function createChecker(program: Program): Checker {
...
-    mapper: TypeMapper | undefined,
-    inheritedProps?: ModelProperty[]
+    mapper: TypeMapper | undefined
   ) {
     for (const prop of node.properties!) {
       if ("id" in prop) {
         const newProp = checkModelProperty(prop, mapper, parentModel);
         checkPropertyCompatibleWithIndexer(parentModel, newProp, prop);
-        defineProperty(properties, newProp, inheritedProps);
+        defineProperty(properties, newProp, parentModel);
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts]:
```suggestion
        defineProperty(properties, newProp);
```

This was a mistake by me, fixing.

File: packages/compiler/core/checker.ts [hunk 2]
```diff
@@ -2136,10 +2136,8 @@ export function createChecker(program: Program): Checker {
       type.namespace?.models.set(type.name, type);
     }
 
-    const inheritedProps = Array.from(walkPropertiesInherited(type));
-
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:2140]: You had the right idea, not using walkPropertiesInherited for this. It's cleaner, IMHO to search up for the particular overridden property. But walkPropertiesInherited did still need fixing for the other use cases where it's warranted.

File: packages/compiler/core/checker.ts [hunk 3]
```diff
@@ -2282,7 +2275,9 @@ export function createChecker(program: Program): Checker {
           })
         );
         return;
-      } else if (!isAssignable) {
+      }
+
+      if (!isAssignable) {
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:2280]: just a style nit here.

File: packages/compiler/core/checker.ts [hunk 4]
```diff
@@ -4586,21 +4582,22 @@ export function getEffectiveModelType(
...
     //
-    // NOTE: Once property overrides are allowed, this code will need to
-    // be updated to check that the current property is not overridden by
-    // the derived type before adding it here. An override would invalidate
-    // this substitution.
+    // Note: We depend on the order of that spread and intersect source
+    // properties here, which is that we see properties sourced from derived
+    // types before properties sourced from their base types.
     addDerivedModels(sources, candidates);
-    addDerivedModels(candidates, sources);
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:4598]: The second call was unnecessary due to the ordering which is also needed to handle overrides correctly. Turns out the only thing wrong was countInheritedProperties, but I added a test and reworded the comment that thought we'd need more changes once overriding happened.

File: packages/compiler/core/checker.ts [hunk 5]
```diff
@@ -4586,21 +4582,22 @@ export function getEffectiveModelType(
...
 
-    // remove candidates that are not common to this property
-    for (const element of candidates) {
-      if (!sources.has(element)) {
-        candidates.delete(element);
+    // remove candidates that are not common to this property.
+    for (const candidate of candidates) {
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:4598]: Fixing my own confusing naming here.

File: packages/compiler/core/checker.ts [hunk 6]
```diff
@@ -4674,33 +4671,54 @@ export function filterModelProperties(
   return finishTypeForProgram(program, newModel);
 }
 
+/**
+ * Gets the property from the nearest base type that is overridden by the
+ * given property, if any.
+ */
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:4677]: I figured this would be generally useful if you really need to get overriden properties so I documented it and exported it. I think you had something almost identical in an earlier revision.

File: packages/compiler/core/checker.ts [hunk 7]
```diff
@@ -4674,33 +4671,54 @@ export function filterModelProperties(
...
+  return undefined;
+}
+
+/**
+ * Enumerates the properties declared by model or inherited from its base.
+ *
+ * Properties declared by more derived types are enumerated before properties
+ * of less derived types.
+ *
+ * Properties that are overridden are not enumerated.
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:4700]: Adding documentation for walkInheritedProperties to cover the ambiguity around what it does in the presence of overrides now.

File: packages/compiler/core/checker.ts [hunk 8]
```diff
@@ -4674,33 +4671,54 @@ export function filterModelProperties(
...
-        count++;
-      }
-    }
-  } else {
-    for (let m: Model | undefined = model; m; m = m.baseModel) {
-      count += m.properties.size;
+  for (const property of walkPropertiesInherited(model)) {
+    if (!filter || filter(property)) {
+      count++;
     }
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:4722]: We can't keep the fast path for the case with no filter any more since adding up the count of each level is wrong when there are overrides.

File: packages/compiler/core/checker.ts [hunk 9]
```diff
@@ -4674,33 +4671,54 @@ export function filterModelProperties(
...
+    for (const property of current.properties.values()) {
+      if (returned.has(property.name)) {
+        // skip properties that have been overridden
+        continue;
       }
+      returned.add(property.name);
+      yield property;
     }
-    yield* returnProps;
-    current = current.baseModel;
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:4689]: Cleaned this up a bit to use the same pattern for looping through bases and avoid making intermediate arrays.

File: packages/compiler/test/parser.test.ts [hunk 10]
```diff
@@ -137,7 +137,6 @@ describe("compiler: parser", () => {
       ["model bar<a, b> = a | b;", [/'{' expected/]],
     ]);
   });
-
```
Comment (nguerrera) [inline packages/compiler/test/parser.test.ts:140]: Reverts an unrelated change to parser.test.ts earlier in the PR. I don't actually care about this, it just fell out of how I was reviewing the overall diff of PR + my changes vs. main.

**Review by tjprescott — APPROVED** [review]

**Comment by tjprescott** [comment]
Thank you for contributing this @nguerrera !!
